expression/grouping_sets.go

```diff
@@ -396,3 +399,203 @@ func (g GroupingExprs) MemoryUsage() int64 {
+//
+// Eg: in the group by a,a,a,a,a with rollup cases. We don't really need rollup
+// the grouping sets as: {a,a,a,a,a},{a,a,a,a},{a,a,a},{a,a},{a},{}, although it
+// can work with copying the column 'a' multi times and distinguish them as a1, a2...
+//
+// since group by {a,a} is same as group by {a,a,a} and {a}...; or since group by
+// {b,a,a} is same as group by {b,a} and {b,a,a,a}. In order to avoid unnecessary
+// redundant column copy, we introduce a new gen_col named as "grouping_pos" to
+// distinguish several grouping set with same basic distinct group items.
+//
+// the above case can be simplified as: {a, gid, gpos}, and gid is used to distinguish
+// deduplicated grouping sets, and gpos is indicated to distinguish duplicated grouping
+// sets. the grouping sets are:
+// {a,0,0}, {a,0,1}, {a,0,2}, {a,0,3}, {a,0,4},{null,1,5}
+//
+// when a is existed, the gid is all the same as 0, that's to say, the first 5 grouping
+// set are all the same one grouping set logically, but why we still need to store them
+// 5 times? because we want to output needed aggregation row as required from rollup syntax;
+// so here we use gpos to distinguish them. In sixth grouping set, the 'a' is null, the
+// gid is newed as 1, it means a new grouping set different with those previous ones, for
+// computation convenience, we increment its gpos number as well.
+//
+// so should care about the gpos when doing the data aggregation? Yes we did, in the above
+// case, shuffle keys should also include the grouping_pos, while the grouping function
+// still only care about the gid as before.
+//
+// mysql> select count(1), grouping(a+1) from t group by a+1, 1+a with rollup;
+// original grouping sets are:
+//        1: {a+1,  1+a}    ---> duplicate! proj a+1 let's say as column#1
+//        2: {a+1}
+//        3: {}
+//  distinguish them in projection item:
+//        Expand schema:[a, column#1, gid, gpos]
+//          |     L1 proj: [a, column#1, 0, 0]; L2 proj: [a, column#1, 0, 1]; L3 proj: [a, null, 1, 2]
+//          |
+//          +--- Projection a+1 as column#1
+//
+// +----------+---------------+
+// | count(1) | grouping(a+1) |
+// +----------+---------------+                            a+1,  gid,  gpos
+// |        1 |             0 | ---+-----> grouping set{column#1,  0,   0}
+// |        1 |             0 | ---+
+// |        1 |             0 | ---+-----> grouping set{column#1,  0,   1}
+// |        1 |             0 | ---+
+// |        2 |             1 | ---------> grouping set{column#1,  1,   2}
+// +----------+---------------+
+// 5 rows in set (0.01 sec)
+//
+// Yes, as you see it, tidb should be able to
+// 1: grouping item semantic equivalence check logic
+// 2: matching select list item to group by item (this seems hard)
+
+// RollupGroupingSets roll up the given expressions, and iterate out a slice of grouping set expressions.
+func RollupGroupingSets(rollupExprs []Expression) GroupingSets {
```

Comment:
currently steps:
1: raw grouping items (may contain complicated func, like a+1)
2: detect the deduplication (for example a+1, 1+a, b)
3: get the deduplicated expression list (a+1,b)
4: the distinct expr list should be projected/copied from the original in the underlying-injected projection (project copy the child schema, and append them behind) (xxx, a+1 as column#1, b)
5: get the projected simple group cols: (column#1, b) because only cols have the unique id, which can be used in grouping sets.
6: restore the original gby list according to new group cols and pos map (column#1, column#1, b)
7: construct rollup grouping sets (pre-condition, all elements are col)

It will be used in buildExpand in next PR